zokrates_core/src/proof_system/zexe/mod.rs

```diff
@@ -0,0 +1,306 @@
+                            let wire = match private {
+                                true => cs.new_witness_variable(|| {
+                                    Ok(witness
+                                        .0
+                                        .remove(&var)
+                                        .ok_or(SynthesisError::AssignmentMissing)?
+                                        .into_zexe())
+                                }),
+                                false => cs.new_input_variable(|| {
+                                    Ok(witness
+                                        .0
+                                        .remove(&var)
+                                        .ok_or(SynthesisError::AssignmentMissing)?
+                                        .into_zexe())
+                                }),
+                            }
+                            .unwrap();
+                            (var.clone(), wire)
+                        }),
+                );
+
+                let main = self.main;
+
+                for statement in main.statements {
+                    match statement {
+                        Statement::Constraint(quad, lin) => {
+                            let a = zexe_combination(
+                                quad.left.clone().into_canonical(),
+                                &mut cs,
+                                &mut symbols,
+                                &mut witness,
+                            )?;
+                            let b = zexe_combination(
+                                quad.right.clone().into_canonical(),
+                                &mut cs,
+                                &mut symbols,
+                                &mut witness,
+                            )?;
+                            let c = zexe_combination(
+                                lin.into_canonical(),
+                                &mut cs,
+                                &mut symbols,
+                                &mut witness,
+                            )?;
+
+                            cs.enforce_constraint(a, b, c)?;
+                        }
+                        _ => {}
+                    }
+                }
+
+                Ok(())
+            }
+            ConstraintSystemRef::None => Err(SynthesisError::MissingCS),
+        }
+    }
+}
+
+impl<T: Field + ZexeFieldExtensions> Computation<T> {
+    pub fn prove(self, params: &Parameters<T::ZexeEngine>) -> Proof<T::ZexeEngine> {
+        let rng = &mut rand_0_7::rngs::StdRng::from_entropy();
+
+        let proof = create_random_proof(self.clone(), params, rng).unwrap();
+
+        let pvk = prepare_verifying_key(&params.vk);
+
+        // extract public inputs
+        let public_inputs = self.public_inputs_values();
+
+        assert!(verify_proof(&pvk, &proof, &public_inputs).unwrap());
+
+        proof
+    }
+
+    pub fn public_inputs_values(&self) -> Vec<<T::ZexeEngine as PairingEngine>::Fr> {
```

Comment:
This function comes up a lot and clones the witness. Would make sense to extract it to `ir` or `witness` and avoid clones.